python_modules/libraries/dagster-dbt/dagster_dbt/dbt_project.py

```diff
@@ -0,0 +1,154 @@
+
+    The default generate_cli_args of ("parse", "--quiet") can be overridden using functools.partial
+    or writing another function that calls this one with that argument set.
+    """
+    from .core.resources_v2 import DbtCliResource
+
+    if deploying or using_dagster_dev() or parse_on_load_opt_in():
+        DbtCliResource(project_dir=os.fspath(project.project_dir)).cli(
+            generate_cli_args,
+            target_path=project.target_dir,
+        ).wait()
+
+
+@experimental
+class DbtProject(DagsterModel):
+    project_dir: Path
+    target_dir: Path
+    manifest_path: Path
+    packaged_project_dir: Optional[Path]
+    manifest_init_fn: ManifestInitFn
+
+    def __init__(
+        self,
+        project_dir: Union[Path, str],
+        *,
+        target_dir: Union[Path, str] = "target",
+        packaged_project_dir: Optional[Union[Path, str]] = None,
+        manifest_init_fn: Optional[ManifestInitFn] = dagster_dbt_manifest_init,
+    ):
+        """Representation of a dbt project.
+
+        Args:
+            project_path (Union[str, Path]):
+                The directory of the dbt project.
+            target_dir (Union[str, Path]):
+                The folder in the project directory to output artifacts.
+                Default: "target"
+            manifest_init_fn (Optional[ManifestInitFn]):
+                A function invoked at init time and via prepare_for_deployment
+                that can be used to ensure the manifest.json is in the right state at
+                the right times.
+                Default: dagster_dbt_manifest_init
+            packaged_project_dir (Optional[Union[str, Path]]):
+                A directory that will contain a copy of the dbt project and the manifest.json
+                when the artifacts have been built. The prepare method will handle syncing
+                the project_path to this directory.
+                This is useful when the dbt project needs to be part of the python package data
+                like when deploying using PEX.
+        """
+        project_dir = Path(project_dir)
+        if not project_dir.exists():
+            raise DagsterDbtProjectNotFoundError(f"project_dir {project_dir} does not exist.")
+
+        packaged_project_dir = Path(packaged_project_dir) if packaged_project_dir else None
+        if using_dagster_dev() and packaged_project_dir:
+            current_project_dir = packaged_project_dir
+        else:
+            current_project_dir = project_dir
+
+        manifest_path = current_project_dir.joinpath(target_dir, "manifest.json")
+
+        super().__init__(
+            project_dir=project_dir,
+            target_dir=target_dir,
+            manifest_path=manifest_path,
+            packaged_project_dir=packaged_project_dir,
+            manifest_init_fn=manifest_init_fn,
+        )
+        if manifest_init_fn:
+            manifest_init_fn(self, deploying=False)
+
+
+def prepare_for_deployment(project: DbtProject):
```

Comment:
hows this name as a free standing function?

Reply:
If it only takes a `project` as a parameter, I feel like putting it on the `DbtProject` object will make things so much simpler for the user.

If we're afraid of making the user subclass `DbtProject` to get custom `prepare_for_deployment` behavior, then we could just instead enforce a pattern of composition rather than subclassing for them to get their desired behavior.

```python
project: DbtProject = ...

# Normal case
project.prepare_for_deployment()

# Composition case
... # Do things before preparing
project.prepare_for_deployment()
...# Do things after

# Complete customization
# 1/ Either they can subclass if they want
# 2/ They can completely eject from using the `.prepare_for_deployment()` function
```